File: leetcode/no0589_N-ary_Tree_Preorder_Traversal.js
/**
 * // Definition for a Node.
 * function Node(val, children) {
 *    this.val = val;
 *    this.children = children;
 * };
 */

/**
 * @param {Node|null} root
 * @return {number[]}
 */
var preorder = function (root) {

    let result = [];

    // Use depth-first search to traverse the binary tree 
    // and record every node encountered.
    function traversal(node) {

        if (node == null) {
            return;
        }

        // Here is the point.
        result.push(node.val);

        for (let i = 0; i < node.children.length; i++) {
            let child = node.children[i];
            traversal(child);
        }
    }
    traversal(root);
    return result;
};

File: kotlin/src/main/kotlin/no0589_Nary_Tree_Preorder_Traversal.kt
class Node(var `val`: Int) {
    var children: List<Node?> = listOf()
}

fun main() {

    /**
     * Definition for a Node.
     * class Node(var `val`: Int) {
     *     var children: List<Node?> = listOf()
     * }
     */
    class Solution {
        fun preorder(root: Node?): List<Int> {
            val result = arrayListOf<Int>();

            // Use depth-first search to traverse the binary tree
            // and record every node encountered.
            fun traversal(node: Node?) {

                if (node == null) {
                    return;
                }

                // Here is the point.
                result += node.`val`

                node.children.forEachIndexed { _, child ->
                    traversal(child);
                }
            }
            traversal(root);
            return result;
        }
    }
}
